Interview crud branch

daido1976 opened the pull request:
#やりたかったこと
---
- 面接日程の一覧表示、新規作成
- 作成した面接日程の編集、削除

#やったこと
----
- pry-rails gem の導入
- i18nの導入
- 面接日程一覧ページの追加
- 面接日程新規作成、編集ページの追加
- 面接日程CRUD操作の実装
- ビューの修正など

#動作確認方法
---
- [1] ユーザーを新規で作成(sign-up)する
- [2] ナビゲーションバーより面接日程一覧ページ(Interviews)にいく
- [3] 下部の新規作成(new interview)ボタンを押す
- [4] 日程を入力し作成(create)ボタンを押す
- [5] 面接日程一覧ページに自分の入力した情報が反映されているのを確認
- [6] 編集(edit)ボタンを押し、日程を更新
- [7] 削除(delete)ボタンを押し、日程を削除

#その他
---
日本語表示と英語表示をいい感じにmixさせたくi18nを使ってみたのですが、結果ja.yml内が散らかってしまいました。
後から修正点に気づくことが多く、コミットが汚くなってしまい申し訳ありません。
「もっとこう書いた方がいいよ！」みたいなところがあればドンドンご指摘ください！
お手数おかけしますが、よろしくお願い致します！

Commit:
Rails g model Interview

Commit:
user interviews association

Commit:
Rails db:migrate

Commit:
Routing&rails g controller interviews

Commit:
Interviews index

Commit:
Interview new

Commit:
interview create

&install Pry-rails

Commit:
Interview edit

Commit:
Interview destroy

Commit:
i18n

どうしてもinterview_datetimeに曜日を表示させたかったのですが、i18nで表示させる方法が分からず、ゴリ押しで書いてしまいました
。
また、deviseの英語表示を使おうと思ったらja.yml内が散らかってしまいました。

Commit:
before_action :set_user in interviewscontroller

tmd45 commented:
i18n さっそく導入されたんですね ✨ 

> 日本語表示と英語表示をいい感じにmixさせたくi18nを使ってみたのですが、結果ja.yml内が散らかってしまいました

> deviseの英語表示を使おうと思ったらja.yml内が散らかってしまいました

i18n のデータは `config/locale/ja.yml` という１ファイルに入れなければいけないものではないです！ 
ですので、たとえば `config/locale/devise_ja.yml` のように、利用する箇所で YAML ファイルを分割することも可能です。

詳しくは RailsGuide などを参照されると良いかと思います 🔖

tmd45 commented on config/routes.rb:
```diff
@@ -1,4 +1,8 @@
+  get 'interviews/index'
+
+  get 'interviews/new'
+
```
下で定義している `resources :interviews` で作られる routing は使えませんか？
あるいは逆に `get 'interviews/index'` で用意した routing は、本当にアプリ内で利用されていますか？
この２点確認してみてください 😎 

実際にどのような routing が登録されたか確認するには `$ bin/rails routes` というコマンドや、
開発環境であれば `http://localhost:3000/rails/info/routes` といった URL で確認できます！

また RailsGuide にも routing に関する説明があるので、改めてチェックしてみると良いかもしれません！

daido1976 replied:
自分で書いた覚えのないルーティングだと思ったらinrterviewコントローラー作成時に「raiils g controller index new」と打った時に生成されたものでした！確認不足で失礼しました。削除しました！

tmd45 commented:
> [4] 日程を入力し作成(create)ボタンを押す

作成ページに入ると、現在の日時がデフォルトで入るようですが
ここで表示されている "時刻" は日本時間（JST）ではないように見えます。

<img width="247" alt="e-navigator" src="https://user-images.githubusercontent.com/294443/36522810-f36d7154-17e1-11e8-99b1-53d167005300.png">

日付・時刻を扱う場合は timezone を気にしてみて下さい！
表示するとき、コード内で扱うとき、データベースに保存されるときとそれぞれ眺めてみるとよいです。

もしそれぞれで扱いたい timezone が異なるときは、
コード ⇔ データベースで扱いやすい timezone で統一し、View に表示するときに「表示したい」timezone に変換するというのが良いかと思います。

tmd45 commented:
i18n 関連でもう１点

> どうしてもinterview_datetimeに曜日を表示させたかったのですが、i18nで表示させる方法が分からず、ゴリ押しで書いてしまいました

i18n には、辞書にある単語をどの言語で表示するか制御する `translate` （View 内で `t` と書けるメソッド）と、もう一つ `localize` というメソッドがあります。
…というところから、View に曜日を表示できるヒントになるかもしれません！

tmd45 commented:
routing でもう１点チェックしてもらいたい部分があります。

以下のように、ユーザ登録した私のユーザ ID は `5` で、こちらのページ遷移は問題なく見えます。

![e-navigator](https://user-images.githubusercontent.com/294443/36523512-3909831c-17e5-11e8-8f23-c2552325dd0e.png)

ここから URL を直接操作すると、ログインしている私以外のユーザの Interviews が見られますが
その他者の Interviews ページの Edit ボタンの URL が、以下のようになっていました。

![e-navigator](https://user-images.githubusercontent.com/294443/36523638-f20fcc0e-17e5-11e8-851b-3aa3380bed77.png)

Edit, Delete のリンクについて確認してみてください！

また、この動作で ログイン中のユーザ（id: 5）は他人（id: 6）の Interviews が閲覧できますし、
もしかしたら他人の面談日程を変更・削除できてしまうかもしれません…
という点についてもチェックしてみてください！ 😎

Commit:
After code review

1.config/locale/devise_ja.ymlの追加
2.不要なroutingの削除
3. timezoneを”Tokyo”に設定
4.interview_datetimeの曜日をlocalizeメソッドで表示
5.controller内にcorrect_userメソッドを定義し、他人のInterviewsを閲覧できないように設定
(編集削除ボタンのみviewから消すというのも考えましたが、一旦こちらで実装してます)

daido1976 commented:
@tmd45 
お忙しい中、今回も丁寧にレビューいただきありがとうございます！
以下修正しましたのでお手すきでご確認いただけますと幸いです。

1.config/locale/devise_ja.ymlの追加
2.不要なroutingの削除
3. timezoneを”Tokyo”に設定
4.interview_datetimeの曜日をlocalizeメソッドで表示
5.controller内にcorrect_userメソッドを定義し、他人のInterviewsを閲覧できないように設定
(編集削除ボタンのみviewから消すというのも考えましたが、一旦こちらで実装してます)

tmd45 commented on app/views/interviews/index.html.erb:
```diff
@@ -11,10 +11,10 @@
     <% @interviews.each do |interview| %>
       <tr>
-        <td><%= interview.interview_datetime.strftime("%Y/%m/%d(#{%w(日 月 火 水 木 金 土)[interview.interview_datetime.wday]}) %H:%M") %></td>
+        <td><%= l interview.interview_datetime, format: :full_date %></td>
```
すっきりしましたね！ ✨

daido1976 replied:
@tmd45 
ご確認ありがとうございます！コミットの分割についても承知しました！訓練していきますね！
引き続きよろしくお願い致します！

tmd45 reviewed (approved):
修正おつかれさまでした！
それぞれバッチリ対応できてると思います 👍 

コミットメッセージに５つのやったことが含まれているので、これは別々のコミットで作業できるとよいですね。

気づくと思いついたところ／気になったところから手をつけ始めてしまったりするのですが、これをやらないようにするのは慣れ（訓練?）だったりするので数をこなしていきましょう 💪 

こちらの PullRequest はマージして、次へ進みましょう :tada:
